naming/pubsub.md

```diff
@@ -0,0 +1,95 @@
+
+Authors:
+
+  - Adin Schmahmann ([@aschmahmann](https://github.com/aschmahmann))
+  
+Reviewers:
+
+-----
+  
+# Abstract
+
+[Inter-Planetary Naming System (IPNS)](/README.md) is a naming system responsible for the creating, reading and updating of mutable pointers to data.
+IPNS consists of a pubic/private asymmetric cryptographic key pair, a record type and a protocol.
+Part of the protocol involves a routing layer that is used for the distribution and discovery of new or updated IPNS records.
+
+The IPNS PubSub router uses as a base [libp2p PubSub](https://github.com/libp2p/specs/tree/master/pubsub), but layers on it persistence so as to ensure IPNS updates are always available to a connected network.
+An inherent property of the IPNS PubSub Router is that IPNS records are republishable by peers other than the peer that originated the record.
+This implies that as long as a peer on the network has an IPNS record it can be made available to other peers (although the records may be ignored if they are received after the IPNS record's End-of-Life/EOL).
+
+# Organization of this document
+
+  - [Introduction](#introduction)
+  - [Protocol](#protocol)
+  - [Overview](#overview)
+  - [API Spec](#api-spec)
+  - [Integration with IPFS](#integration-with-ipfs)
+
+# Introduction
+
+Each time a node publishes an updated IPNS record for a particular key it is propagated by the router into the network where network nodes can choose to accept or reject the new record.
+When a node attempts to retrieve an IPNS record from the network it uses the router to query for the IPNS record(s) associated with the IPNS key; the node then validates the received records.
+
+In this spec we address building a router based on a PubSub system, particularly focusing on libp2p PubSub.
+
+# PubSub Protocol Overview
+
+The protocol has four components:
+- [IPNS Records and Validation](/README.md)
+- [libp2p PubSub](https://github.com/libp2p/specs/tree/master/pubsub)
+- Translating an IPNS record name to/from a PubSub topic
+- Layering persistence onto libp2p PubSub
+
+# Translating an IPNS record name to/from a PubSub topic
+
+For a given IPNS local record key described in the IPNS Specification the PubSub topic is:
+
+**Topic format:** `/record/base64url-unpadded(key)`
+
+where base64url-unpadded is an unpadded base64url as specified in [IETF RFC 4648](https://tools.ietf.org/html/rfc4648)
+
+# Layering persistence onto libp2p PubSub
+
+libp2p PubSub does not have any notion of persistent data built into it. However, we can layer persistence on top of PubSub by utilizing [libp2p Fetch](https://github.com/libp2p/specs/).
+
+The protocol has the following steps:
+1. Start State: Node `A` subscribes to the PubSub topic `t` corresponding to the local IPNS record key `k`
+2. `A` notices that a node `B` has connected to it and subscribed to `t`
+3. Some time passes (might be 0 seconds, or could use a more complex system to determine the duration)

```

Comment:
It's the other way, `Fetch` is a pull protocol so when `B` joins it will send out mass `Fetch`s to nodes as they connect to it. However, it's not like this is a heavy protocol and all the nodes have already established connections with us and sent us `Subscribe` RPCs over PubSub.

If an IPNS-over-PubSub wants to have a handler-pool to throttle, or wants to wait some amount of time based on the last `Fetch` they sent that's absolutely acceptable. I don't think 0 is too problematic though.

Reply:
Got it, thanks for explanation!

Reply:
Referenced your concern in the go implementation [PR](https://github.com/libp2p/go-libp2p-pubsub-router/pull/33/files#r314280601)

If you have more comments to add, take a look there since this is probably more of an implementation specific discussion then spec related.